src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -108,20 +122,80 @@ Score(const size_t queryIndex, TreeType& referenceNode)
 
     // Calculate kernel value based on reference node centroid.
     if (tree::TreeTraits<TreeType>::FirstPointIsCentroid)
-    {
       kernelValue = EvaluateKernel(queryIndex, referenceNode.Point(0));
-    }
     else
-    {
-      kde::KDEStat& referenceStat = referenceNode.Stat();
       kernelValue = EvaluateKernel(queryPoint, referenceStat.Centroid());
-    }
 
     densities(queryIndex) += referenceNode.NumDescendants() * kernelValue;
 
     // Don't explore this tree branch.
     score = DBL_MAX;
   }
+  else if (monteCarlo &&
+           referenceNode.NumDescendants() >= MCAccessCoef * initialSampleSize &&
+           std::is_same<KernelType, kernel::GaussianKernel>::value)
+  {
+    // Calculate reference node depth.
+    if (referenceStat.Depth() == 0)
+      referenceStat.Depth() = CalculateDepth(referenceNode);
+
+    // Monte Carlo probabilistic estimation.
+    const double currentAlpha =
+        MCBeta / std::pow(2, referenceNode.Stat().Depth() + 1);
```

Comment:
What do you think about the idea of "reclaiming" error that Dongryeol mentions in the paper?  Honestly I suspect that is responsible for the largest part of the speedup.  The idea is detailed more in this paper: https://arxiv.org/pdf/1206.6857.pdf in Section 5 (see under Theorem 2).

Reply:
I think I made this work (at least for single-trees) but at the moment I can't see that much of a difference in tests so I'm testing whether it's actually working or not :)

Reply:
Implemented for single-trees in a68d28c8e3ac9ab949eda76469b665104d32912c.
After some tests I think it's robust. Although results depends very much on input parameters, it performs ~25-30% less base cases (compared to the non-reclaiming version) and the amount of estimations out of the relative error bound remain the same :)